presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -152,13 +160,16 @@ public static Query create(
     {
         Query result = new Query(sessionContext, query, queryManager, sessionPropertyManager, exchangeClient, dataProcessorExecutor, timeoutExecutor, blockEncodingSerde);
 
-        result.queryManager.addOutputInfoListener(result.getQueryId(), result::setQueryOutputInfo);
+        // register listeners after submission finishes
+        addSuccessCallback(result.submissionFuture, () -> {
+            result.queryManager.addOutputInfoListener(result.getQueryId(), result::setQueryOutputInfo);
```

Comment:
There might be a race condition here where the query can start and add output locations before we register the listener and thus we miss the notification